src/Microsoft.ML.Data/Transforms/ValueToKeyMappingTransformer.cs

```diff
@@ -295,12 +248,12 @@ private ColInfo[] CreateInfos(Schema inputSchema)
         }
 
         internal ValueToKeyMappingTransformer(IHostEnvironment env, IDataView input,
```

Comment:
>ValueToKeyMappingTransformer [](start = 17, length = 28)

do we want to keep the ctors of the transformers public, symmetrically with the model parameters (former predictors)?

Reply:
This might be a bad example, since we've determined after some time that having public constructors for model parameters is pointless (see e.g., #1866 and leadup discussion to this, also cc @eerhardt, @glebuk). That said, in favor of your point, direct construction of transformers *is* potentially useful -- unlike direct creation of model parameters, with which by itself literally nothing can be done.

However, unless we have a solid immediate scenario that literally cannot be accommodated using the usual `MLContext` pathway (not the case here), we should not do it. I sort of like APIs with one way for a user to do things, not multiple, at least for v1. (And even if we did have such a scenario, considering the timeframes, we probably wouldn't want to do it anyway.)

Reply:
As per #1098, we do want to make them `internal`.  For V1, MLContext would be the way to create  them all.

Regarding model parameters, there is a separate issue for that #1866  .  We can  do those as a separate effort / PR perhaps ?

---
In reply to: [252551812](https://github.com/dotnet/machinelearning/pull/2340#discussion_r252551812) [](ancestors = 252551812)

Reply:
I don't think we should couple public-ness of transformers with model params.  Those are different things.  As Rogan suggested in #1886, it can be valuable to have public model ctors for a variety of scenarios.  But that is  a set of different reason vs transformers.  However, in case of transformers, I agree, if there is a way to create them with MLContext, there is no burning reason to have a public ctor.

Reply:
Background:
- Trainable transformers can only be instantiated through estimators (using .Fit() )
- Non trainable transforms can be instantiated directly (through their constructor)
- We only create estimators through MLContext, not transformers

I think we should keep the constructors of the non trainable transformers public, while internalizing the constructors of the estimators.

Reply:
@artidoro .  Perhaps we should also make the constructors for non-trainable  transforms `internal`.

I believe @tom and @gleb  also mention the same in their comments. 


---
In reply to: [252782663](https://github.com/dotnet/machinelearning/pull/2340#discussion_r252782663) [](ancestors = 252782663)

Reply:
After discussing with Tom and Gleb, I think the decision is to internalize constructors for all transformers. They will only be instantiated through their respective estimators.

The question remains on how to make it look less silly when you have a non trainable transformer that needs to be instantiated using .Fit() on some data.